### PR TITLE
Issue 8043: Client application stuck at initiateCheckpoint Api call

### LINKED ISSUE #7215: Client application stuck at initiateCheckpoint Api call.
**Problem description**
Client application is stuck at initiateCheckpoint() Api call when it uses Pravega internal thread pool to process checkpointing task as
single thread is hardcoded as part of SocketConnectionFactoryImpl while creating new instance of ReaderGroupManagerImpl.

**Problem location**
https://github.com/pravega/pravega/blob/1825cad1b14e39dc9a869a4b38fef8fa62c3e046/client/src/main/java/io/pravega/client/admin/ReaderGroupManager.java#L62C75-L62C102

**Suggestions for an improvement**
Number of threads for SocketConnectionFactoryImpl  should be more than 1, 
Write a testcase for validating minimum number of required threads.

### PR DESCRIPTION
**Pull request by ShwethaSNayak**
**Change log description**  
Client application get stuck when it triggers initiateCheckpoint API with Pravega internal executorService.
ReaderGroupManager constructor takes thread count as 1 while constructing SocketClientFactory, because of which initiatecheckpoint call get stuck.

**Purpose of the change**  
closes #7215 

**What the code does**  
Change thread count to 3 in SocketClientFactory while constructing ReaderGroupManager. 

**How to verify it**  
System test added for this change.
In system test added assertion for multiple checkpoint completion stage to make sure checkpoint is not blocked any more

### CODE REVIEW

File: test/system/src/test/java/io/pravega/test/system/ReaderCheckpointWithInternalThreadpool.java [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+                }
+            } while (event.isCheckpoint() && event.getCheckpointName() != CHECKPOINT2);
+            //stop reading if CHECKPOINT_2 is acked by reader.
+            log.info("No more events from {}/{} for readerId: {}", scope, STREAM, readerId);
+        } //reader.close() will automatically invoke ReaderGroup#readerOffline(String, Position)
+        return null;
+    }
+
+    private URI fetchControllerURI() {
+        Service conService = Utils.createPravegaControllerService(null);
```
Comment (tkaitchuck) [inline test/system/src/test/java/io/pravega/test/system/ReaderCheckpointWithInternalThreadpool.java:241]: This is creating a controller service. Probably not what is intended. (As the controller was already started on line 100)
This appears to be copied from somewhere, can you make sure to correct this in all the files.

File: test/system/src/test/java/io/pravega/test/system/ReaderCheckpointWithInternalThreadpool.java [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+        URI controllerUri = ensureControllerRunning(zkUri);
+        ensureSegmentStoreRunning(zkUri, controllerUri);
+    }
+
+    @Before
+    public void setup() {
+        controllerURI = fetchControllerURI();
+        streamManager = StreamManager.create(Utils.buildClientConfig(controllerURI));
+
+        connectionFactory = new SocketConnectionFactoryImpl(ClientConfig.builder().build());
```
Comment (tkaitchuck) [inline test/system/src/test/java/io/pravega/test/system/ReaderCheckpointWithInternalThreadpool.java:109]: Here two different client configs are being used. And a duplicate one is also created on 111

File: test/system/src/test/java/io/pravega/test/system/ReaderCheckpointWithInternalThreadpool.java (hunk 1 above)
Comment (ShwethaSNayak) [reply test/system/src/test/java/io/pravega/test/system/ReaderCheckpointWithInternalThreadpool.java:241]: Thats right, used existing precondition logic, let me correct that

File: test/system/src/test/java/io/pravega/test/system/ReaderCheckpointWithInternalThreadpool.java (hunk 2 above)
Comment (ShwethaSNayak) [reply test/system/src/test/java/io/pravega/test/system/ReaderCheckpointWithInternalThreadpool.java]: done

**Comment by ShwethaSNayak** [comment]
System test run with thread count as 1 :
https://asd-nautilus-jenkins.isus.emc.com/job/pravega-development-system-tests-2/2510/

System test run with thread count as 3 : 
https://asd-nautilus-jenkins.isus.emc.com/job/pravega-development-system-tests-2/2511/

File: test/system/src/test/java/io/pravega/test/system/ReaderCheckpointWithInternalThreadpool.java (hunk 1 above)
Comment (ShwethaSNayak) [reply test/system/src/test/java/io/pravega/test/system/ReaderCheckpointWithInternalThreadpool.java:234]: Its actually creating controller service object (not starting new controller service instance) just to fetch URI from service.

**Comment by ShwethaSNayak** [comment]
Post pulling latest merge from Master branch I do see few failures in tests, will correct that

**Review by tkaitchuck — APPROVED** [review]

File: test/system/src/test/java/io/pravega/test/system/ReaderCheckpointWithInternalThreadpool.java [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+        } catch (Exception e)  {
+            if (e instanceof InterruptedException || e instanceof ExecutionException || e instanceof TimeoutException) {
+                log.error("Exception while initiating checkpoint: {}", checkPointName, e);
+                fail("Exception while initiating checkpoint is not expected");
+            }
+        }
+        return result;
+    }
+
+    private void initiateCheckpoointAndVerify(final ReaderGroup readerGroup) {
```
Comment (a6dulaleem) [inline test/system/src/test/java/io/pravega/test/system/ReaderCheckpointWithInternalThreadpool.java:172]: There is one typo, Could you please change the method name to `initiateCheckpointAndVerify`?

File: test/system/src/test/java/io/pravega/test/system/ReaderCheckpointWithInternalThreadpool.java [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+        ReaderGroup readerGroup = readerGroupManager.getReaderGroup(READER_GROUP_NAME);
+
+        // Write events
+        int start = 1;
+        int end = 500;
+        log.info("Write events with range [{},{})", start, end);
+        writeEventsToStream(SCOPE, IntStream.range(start, end).boxed().collect(Collectors.toList()));
+        asyncReadEvents(SCOPE, READER);
+        //Initiate 2 checkpoints and make sure its completed
+        initiateCheckpoointAndVerify(readerGroup);
```
Comment (a6dulaleem) [inline test/system/src/test/java/io/pravega/test/system/ReaderCheckpointWithInternalThreadpool.java:152]: Same Here ?

File: test/system/src/test/java/io/pravega/test/system/ReaderCheckpointWithInternalThreadpool.java (hunk 1 above)
Comment (a6dulaleem) [reply test/system/src/test/java/io/pravega/test/system/ReaderCheckpointWithInternalThreadpool.java:234]: I am still confused, on line number 101 we are already getting the controller URI we can directly assign that to controllerURI variable and we can use that. 

in that case we don't need to create this method.

File: test/system/src/test/java/io/pravega/test/system/ReaderCheckpointWithInternalThreadpool.java (hunk 4 above)
Comment (ShwethaSNayak) [reply test/system/src/test/java/io/pravega/test/system/ReaderCheckpointWithInternalThreadpool.java]: Done

File: test/system/src/test/java/io/pravega/test/system/ReaderCheckpointWithInternalThreadpool.java (hunk 3 above)
Comment (ShwethaSNayak) [reply test/system/src/test/java/io/pravega/test/system/ReaderCheckpointWithInternalThreadpool.java]: Done

**Review by a6dulaleem — APPROVED** [review]
LGTM

**Review by abhinb — APPROVED** [review]

**Pull request merged**